rfcs/20190815-tfx-notebook.md

```diff
@@ -0,0 +1,563 @@
+# Path which can be listened to by the model server.  Pusher will output the
+# trained model here.
+_serving_model_dir = os.path.join(_output_bucket, 'serving_model',
+                                  _pipeline_name)
+
+# Region to use for Dataflow jobs and AI Platform training jobs.
+#   Dataflow: https://cloud.google.com/dataflow/docs/concepts/regional-endpoints
+#   AI Platform: https://cloud.google.com/ml-engine/docs/tensorflow/regions
+_gcp_region = 'us-central1'
+
+# A dict which contains the training job parameters to be passed to Google
+# Cloud AI Platform. For the full set of parameters supported by Google Cloud AI
+# Platform, refer to
+# https://cloud.google.com/ml-engine/reference/rest/v1/projects.jobs#Job
+_ai_platform_training_args = {...}
+
+# A dict which contains the serving job parameters to be passed to Google
+# Cloud AI Platform. For the full set of parameters supported by Google Cloud AI
+# Platform, refer to
+# https://cloud.google.com/ml-engine/reference/rest/v1/projects.models
+_ai_platform_serving_args = {...}
+
+# Beam args to run data processing on DataflowRunner.
+_beam_pipeline_args = [...]
+
+# The rate at which to sample rows from the Chicago Taxi dataset using BigQuery.
+# The full taxi dataset is > 120M record.  In the interest of resource
+# savings and time, we've set the default for this example to be much smaller.
+# Feel free to crank it up and process the full dataset!
+_query_sample_rate = 0.001  # Generate a 0.1% random sample.
+
+# This is the upper bound of FARM_FINGERPRINT in Bigquery (ie the max value of
+# signed int64).
+_max_int64 = '0x7FFFFFFFFFFFFFFF'
+
+# The query that extracts the examples from BigQuery.  The Chicago Taxi dataset
+# used for this example is a public dataset available on Google AI Platform.
+_query = ...
+##############################################################################
+
+# TODO(USER)
+p = pipeline.Pipeline(
+    pipeline_name=,
+    pipeline_root=,
+    components=[
+        example_gen, statistics_gen, infer_schema, validate_stats, transform,
+        trainer, model_analyzer, model_validator, pusher
+    ],
+    additional_pipeline_args={
+        'beam_pipeline_args': beam_pipeline_args,
+        # Optional args:
+        # 'tfx_image': custom docker image to use for components.
+        # This is needed if TFX package is not installed from an RC
+        # or released version.
+    },
+    log_root='/var/tmp/tfx/logs')
+
+kubeflow_pipeline = KubeflowRunner().run(p)
+
+# Export notebook contents.
+context = InteractiveContext()
+# TODO(USER): Name of the notebook file to be used for retrieving
+# notebook contents. IPython kernels are agnostic to notebook metadata by design,
+# and it seems that existing workarounds to retrieve the notebook filename are not
+# universally robust (https://github.com/jupyter/notebook/issues/1000).
+context.export_to_pipeline(notebook_filename='taxi_pipeline_interactive.ipynb',
+                           type='kubeflow')
+```
```

Comment:
What exactly is the benefit of having these templates versus non-notebook code? Can you clarify what the end to end user story is for something like an Airflow pipeline?

Reply:
I think the only difference here would be exposure in the tutorial notebook for this use case.

The end-to-end user story would be (1) in a notebook, a user defines, tries out and iterates on a ML pipeline using TFX components; (2) the user decides that the pipeline is in a state where they are comfortable using it in production and runs `export_to_pipeline(..., type='airflow')` to write the output to a Python pipeline script; (3) the user modifies the pipeline script as necessary (e.g. pointing it to their Airflow cluster) and runs the script to submit / trigger their final production pipeline.

Reply:
It sounds like the export is a key part of this product, and the primary goal in most cases. In that case, it seems like restructuring this to make sure export is easy and flawless seems wise? Why would a user experiment here if they have to do a lot of manual work to translate for export?